### ISSUE TITLE
Support niche optimization for Option<NonZero{Integer}>

### ISSUE THREAD

**Issue by seritools** [rkyv/rkyv#241]
It'd be great to have this provided out-of-the-box. This works because `Option<NonZero{Integer}>` is guaranteed to be the same as that `{Integer}` type.
`rend`'s `LittleEndian`/`BigEndian` wrappers also are repr-transparent, so it'll even work when `archive_le`/`archive_be` is enabled, as zero is the same in both LE and BE, thankfully.

I've got a very rough impl (LE only; somewhat spooky transmutes) going - pretty new to rkyv, so I'm not sure it's entirely correct, but I hope it gets the idea across at least :)

```rs
pub struct NicheOptionNonZeroInteger;

macro_rules! impl_opt_non_zero_int_niche {
    ($non_zero:ty, $raw:ty) => {
        impl ArchiveWith<Option<$non_zero>> for NicheOptionNonZeroInteger {
            type Archived = Option<rkyv::rend::LittleEndian<$non_zero>>;

            type Resolver = Resolver<$raw>;

            unsafe fn resolve_with(
                field: &Option<$non_zero>,
                pos: usize,
                resolver: Self::Resolver,
                out: *mut Self::Archived,
            ) {
                unsafe { std::mem::transmute::<_, $raw>(*field) }.resolve(
                    pos,
                    resolver,
                    out as *mut rkyv::rend::LittleEndian<$raw>,
                );
            }
        }

        impl<S: Fallible + ?Sized> SerializeWith<Option<$non_zero>, S> for NicheOptionNonZeroInteger
        where
            $raw: Serialize<S>,
        {
            fn serialize_with(
                field: &Option<$non_zero>,
                serializer: &mut S,
            ) -> Result<Self::Resolver, S::Error> {
                unsafe { std::mem::transmute::<_, $raw>(*field) }.serialize(serializer)
            }
        }

        impl<D: Fallible + ?Sized>
            DeserializeWith<Option<rkyv::rend::LittleEndian<$non_zero>>, Option<$non_zero>, D>
            for NicheOptionNonZeroInteger
        where
            Archived<$raw>: Deserialize<$raw, D>,
        {
            fn deserialize_with(
                field: &Option<rkyv::rend::LittleEndian<$non_zero>>,
                deserializer: &mut D,
            ) -> Result<Option<$non_zero>, D::Error> {
                Ok(<$non_zero>::new(
                    unsafe { std::mem::transmute::<_, Archived<$raw>>(*field) }
                        .deserialize(deserializer)?,
                ))
            }
        }
    };
}

impl_opt_non_zero_int_niche!(NonZeroU64, u64);
// ...
```

**Comment by djkoloski** [comment]
This seems like it should be pretty easy to support. This will use the [`Niche`](https://docs.rs/rkyv/latest/rkyv/with/struct.Niche.html) wrapper.

**Comment by djkoloski** [comment]
Fixed by 39cdd76. This will be released in rkyv 0.7.30 (in at most a week).

**Comment by seritools** [comment]
Awesome, thank you!
